### PR TITLE
Minor: Improve documentation for registering `AnalyzerRule`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?
Related to https://github.com/apache/arrow-datafusion/issues/9519 

## Rationale for this change

When filing https://github.com/apache/arrow-datafusion/issues/9519 I didn't think DataFusion allowed registering AnalyzerRules. However, in the middle of filing a ticket requesting we do so, I found  [`SessionState::add_analyzer_rule`](https://docs.rs/datafusion/latest/datafusion/execution/context/struct.SessionState.html#method.add_analyzer_rule)

I would like to make this clearer to find in the future

## What changes are included in this PR?

1. Improve doc comments, add hints about how to reguster contexts
 
## Are these changes tested?
Doc tests
## Are there any user-facing changes?
Documentation only (no code changes)

### CODE REVIEW

File: datafusion/core/src/execution/context/mod.rs [hunk 1]
```diff
@@ -1587,7 +1587,7 @@ impl SessionState {
         self
     }
 
-    /// Replace the default query planner
+    /// use `query_planner` to plan queries, replacing the default
```
Comment (comphead) [inline datafusion/core/src/execution/context/mod.rs:1590]:
```suggestion
    /// override default query planner with `query_planner`
```

File: datafusion/core/src/execution/context/mod.rs [hunk 2]
```diff
@@ -1596,7 +1596,7 @@ impl SessionState {
         self
     }
 
-    /// Replace the analyzer rules
+    /// Replace the entire list of [`AnalyzerRule`]s used to optimize plans.
```
Comment (comphead) [inline datafusion/core/src/execution/context/mod.rs:1599]:
```suggestion
    /// Override the [`AnalyzerRule`]s optimizer plan rules.
```

File: datafusion/core/src/execution/context/mod.rs [hunk 3]
```diff
@@ -1641,12 +1643,13 @@ impl SessionState {
         self
     }
 
-    /// Adds a new [`PhysicalOptimizerRule`]
+    /// Add `physical_optimizer_rule` to the end of the list of
+    /// [`PhysicalOptimizerRule`]s used to rewrite queries.
     pub fn add_physical_optimizer_rule(
         mut self,
-        optimizer_rule: Arc<dyn PhysicalOptimizerRule + Send + Sync>,
+        physical_optimizer_rule: Arc<dyn PhysicalOptimizerRule + Send + Sync>,
```
Comment (comphead) [inline datafusion/core/src/execution/context/mod.rs:1650]: 👍

**Review by comphead — APPROVED** [review]
lgtm thanks @alamb I left some suggestion for more concise phrasing if you okay with that

**Review by jayzhan211 — APPROVED** [review]

File: datafusion/core/src/physical_optimizer/optimizer.rs [hunk 4]
```diff
@@ -34,8 +34,12 @@ use crate::physical_optimizer::topk_aggregation::TopKAggregation;
 use crate::{error::Result, physical_plan::ExecutionPlan};
 
 /// `PhysicalOptimizerRule` transforms one ['ExecutionPlan'] into another which
-/// computes the same results, but in a potentially more efficient
-/// way.
+/// computes the same results, but in a potentially more efficient way.
+///
+/// Use [`SessionState::add_physical_optimizer_rule`] to register additional
+/// `PhysicalOptimizerRule`s..
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/optimizer.rs:40]:
```suggestion
/// `PhysicalOptimizerRule`s.
```

**Pull request merged**